qualtran/bloqs/mod_arithmetic/mod_subtraction_test.py

```diff
@@ -119,3 +129,117 @@ def test_notebook():
+    assert counts['n_ccz'] == 6 * n - 3
+
+
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 6)]
+)
+def test_modsub_decomposition(dtype, bitsize, prime):
+    b = ModSub(dtype(bitsize), prime)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+
+
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 6)]
+)
+def test_modsub_bloq_counts(dtype, bitsize, prime):
+    b = ModSub(dtype(bitsize), prime)
+    qlt_testing.assert_equivalent_bloq_counts(b)
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 6)]
+)
+def test_modsub_classical_action(dtype, bitsize, prime):
+    b = ModSub(dtype(bitsize), prime)
+    cb = b.decompose_bloq()
+    for x, y in itertools.product(range(prime), repeat=2):
+        assert b.call_classically(x=x, y=y) == cb.call_classically(x=x, y=y) == (x, (y - x) % prime)
+
+
+def test_modsub_classical_action_fast():
+    bitsize = 10
+    prime = 541
+    rng = np.random.default_rng(13214)
+    b = ModSub(QUInt(bitsize), prime)
+    cb = b.decompose_bloq()
+    for x, y in rng.choice(prime, (10, 2)):
+        assert b.call_classically(x=x, y=y) == cb.call_classically(x=x, y=y) == (x, (y - x) % prime)
+
+
+def test_cmodsub_cost():
+    n, p = sympy.symbols('n p')
+    b = CModSub(QMontgomeryUInt(n), p)
+    counts = get_cost_value(b, QECGatesCost()).total_t_and_ccz_count()
+
+    # Litinski 2023 https://arxiv.org/abs/2306.08585
+    # Figure/Table 8. Lists modular subtraction as 7n toffoli.
+    assert counts['n_t'] == 0
+    assert counts['n_ccz'] == 7 * n - 1
+
+
+@pytest.mark.parametrize('cv', range(2))
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 6)]
+)
+def test_cmodsub_decomposition(cv, dtype, bitsize, prime):
+    b = CModSub(dtype(bitsize), prime, cv)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+
+
+@pytest.mark.parametrize('cv', range(2))
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 6)]
+)
+def test_cmodsub_bloq_counts(cv, dtype, bitsize, prime):
+    b = CModSub(dtype(bitsize), prime, cv)
+    qlt_testing.assert_equivalent_bloq_counts(b)
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('cv', range(2))
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length() - 1, 6)]
+)
+def test_cmodsub_classical_action(cv, dtype, bitsize, prime):
```

Comment:
niiiiiiiice

Reply:
lol ... I just noticed this ... will fix in the next PR